CPS-0019? | Light Client Protocols

cleanerm5 opened the pull request:
Hi,

There are a few activities ongoing that require the development of so called light client protocols for Cardano. The ones that came to my awareness are:
- the IBC implementation of CF (https://github.com/cardano-foundation/cardano-ibc-incubator) 
- the Bitcoin-Cardano bridge proposed by BitcoinOS and with the general approach described in this paper (https://assets-global.website-files.com/661e3b1622f7c56970b07a4c/662a7a89ce097389c876db57_BitSNARK__Grail.pdf) 
- the IOG research work on trust-minimized light client based wallets

There are various ways how to implement light clients on Cardano (whereas because of the core protocol not providing so called state proofs out of the box, there is no canonical light client design available for Cardano at the moment). In that sense, I wanted to create a common point of reference and problem description with this CPS to which concrete light client designs can refer to when they are documented as a CIP.

At least for the IBC implementation we plan to document the Mithril based light client as a CIP and the same would make sense for the respective light client protocol to be developed for a Bitcoin-Cardano bridge (in case the bridge follows a light client based security model), because both designs can be used for other use cases as well.

Any feedback is highly appreciated.

Kind regards

---

([rendered latest version](https://github.com/cleanerm5/CIPs/blob/master/CPS-%3F%3F%3F%3F/README.md))

cleanerm5 committed:
Added draft of light client CPS

cleanerm5 committed:
Merge branch 'cardano-foundation:master' into master

cleanerm5 committed:
Made abstract a bit more lean

cleanerm5 committed:
Merge branch 'master' of github.com:cleanerm5/CIPs

cleanerm5 committed:
Refined most sections

cleanerm5 committed:
Refined further sections

cleanerm5 committed:
Simplified reference

cleanerm5 committed:
added footnote mark down

cleanerm5 committed:
Fixed footnotes

cleanerm5 committed:
Fixed footnotes

rphair commented on CPS-????/README.md:
```diff
@@ -0,0 +1,96 @@
+CPS: ?
+Title: Light Client Protocols for Cardano
+Status: Open
+Category: Tools
```
```suggestion
Category: Network
```
[suggestion is only put forward for debate: not a definite recommendation at this point]

I think it would be a good juncture to consider whether protocols for these _Use Cases_, and defined as in the subsection _Light Client Definition_, might be considered in the `Network` category.  So far I think we've assumed that this category is only intended for the Cardano Node's network behaviour.  I don't recall this coming up in prior discussion & wanted to include other `Network` contributors like @jpraynaud if possible.

rphair replied:
Suggestion was dismissed today (as my misunderstanding) since this doesn't pertain to "network" behaviour but rather methods of [i.e., tools for] integrating other protocol activities with the main protocol.

rphair commented on CPS-????/README.md:
```diff
@@ -0,0 +1,96 @@
+---
+CPS: ?
+Title: Light Client Protocols for Cardano
```
```suggestion
Title: Light Client Protocols
```
Audience for CIP/CPS is entirely focused on Cardano, so the phrase `for Cardano` is not necessary & has generally been removed for other documents in the repository.

rphair commented on CPS-????/README.md:
```diff
@@ -0,0 +1,96 @@
+License: CC-BY-4.0
+---
+
+<!-- Existing categories:
+
+- Meta     | For meta-CIPs which typically serves another category or group of categories.
+- Wallets  | For standardisation across wallets (hardware, full-node or light).
+- Tokens   | About tokens (fungible or non-fungible) and minting policies in general.
+- Metadata | For proposals around metadata (on-chain or off-chain).
+- Tools    | A broad category for ecosystem tools not falling into any other category.
+- Plutus   | Changes or additions to Plutus
+- Ledger   | For proposals regarding the Cardano ledger (including Reward Sharing Schemes)
+- Catalyst | For proposals affecting Project Catalyst / the Jörmungandr project
+
+-->
+
```
```suggestion
```
(remove comment scaffolding)

It seems I've also forgotten to update, in the last CIP-0001 revision, the template files to include & define the `Consensus` and `Network` categories.  I'm leaving this as a "note to self" to fix these files ASAP:
- https://raw.githubusercontent.com/cardano-foundation/CIPs/refs/heads/master/.github/CIP-TEMPLATE.md
- https://raw.githubusercontent.com/cardano-foundation/CIPs/refs/heads/master/.github/CPS-TEMPLATE.md

rphair replied:
(will add `Consensus` & `Network` categories in forthcoming PR)

rphair reviewed (commented):
Thanks, I'm marking `Triage` to begin discussion of this at the next CIP meeting (https://hackmd.io/@cip-editors/102) if not earlier in this thread.

This will be an introduction of the CPS, rather than a full review, but I would also plan to discuss the scope enough to see if a category of `Network` rather than `Tools` would be appropriate.  @cleanerm5 you have obviously put a lot of thought into applications so I would hope you can come to this meeting in a week's time to answer this question first.

rphair committed:
remove comment scaffolding

rphair committed:
removed redundant phrase from title

rphair commented on CPS-????/README.md:
```diff
@@ -0,0 +1,83 @@
+---
+CPS: ?
+Title: Light Client Protocols
```
@cleanerm5 if you could work with reviewers & editors to make or propose a title change, as per the above comment, it would be appreciated before the end-of-year work hiatus begins.

colll78 replied:
It is about light client protocols, but not about the networking, it is about the technological properties of light client proofs to verify state transitions on Cardano in contracts on other chains.

rphair replied:
@colll78 & editors: though I was hoping to hear from @cleanerm5 about possibly choosing a different title — due to the current title inappropriately suggesting the more familiar networking "protocols" to a general audience — I can see what's meant by "protocols" as a means of consensus is clearly defined in the Abstract as well as the relevant academic & practical literature.

Therefore @Ryun1 @Crypto2099 @perturbing I think it's fine from a community perspective to include it unchanged here, though the reader's mileage may vary... resolving.

rphair reviewed (commented):
Confirmed as candidate in CIP meeting today: pending a title revision that reflects _not_ a standard / framework for "light client protocols" but rather a means of incorporating these into the main protocol with proofs (as @colll78 pointed out).

Ordinarily we would wait for a title accurately representing the CIP scope, but given the high profile concurrent work that relates to this CIP we decided to put the number through immediately.

Please @cleanerm5 rename the directory to `CPS-0019` and re-link your original comment to the new pathname of the latest proposal in your branch. 🎉

rphair commented on CPS-????/README.md:
```diff
@@ -0,0 +1,83 @@
+---
+CPS: ?
```
```suggestion
CPS: 142
```

rphair replied:
corrected: see https://github.com/cardano-foundation/CIPs/pull/942/commits/5c37beb945bf120a96441c8412dd34fda4eca093

rphair committed:
assign CIP number 142

rphair committed:
adding initial PR number

rphair committed:
editor (me) accidentally assigned CIP instead of CPS number

... in https://github.com/cardano-foundation/CIPs/pull/942#discussion_r1878681750

EzequielPostan reviewed (commented):
This was a nice reading, thank you for the document.
I may add a comment/question. Is it actually the role of light clients to improve the finality properties of Cardano (or any underlying chain)? I can understand the challenge of finality for application development, but I wanted to ask if the protocols to generate state proofs (as defined in the document) should be mixed with underlying consensus finality management

polinavino commented:
FYI - There is a "Light Client Service" section in the following CIP about nested transactions

The idea is that batching transactions in a specific way, called Nested Transactions (plus some extra versatility for script use), allows users to define an LC protocol which is NOT about the LC querying a service provide for blockchain data, but instead, an LC asks the SP to help them build a transaction that satisfies some properties

https://github.com/cardano-foundation/CIPs/pull/862

rphair reviewed (approved):
@cleanerm5 this was hanging in limbo for a period from the unresolved query in https://github.com/cardano-foundation/CIPs/pull/942#discussion_r1878674513 ... I'm cleaning through similarly "waiting for response" candidates that we can still promote and I think yours was actually ready long ago.

@Ryun1 @Crypto2099 @perturbing it's especially important that we have a proper CPS container at least for Mithril & all others enumerated here which may have CIPs for them.  Though I'm not an expert reviewer I can find no fault with this document nor how it paves the way for development in this field... so marking `Last Check` for next CIP meeting in 2 days' time (https://hackmd.io/@cip-editors/113) to merge unless some fault is identified.

> [!NOTE]
> @cleanerm5 @polinavino if you can consider at the last minute including candidate #862 as suggested in https://github.com/cardano-foundation/CIPs/pull/942#issuecomment-2596404536 then please contribute an editorial suggestion... otherwise we will leave it to a later PR update if relevant.

rphair commented:
@cleanerm5 after feedback from yesterday's CIP meeting I'm putting this back in `Confirmed` status (= active review but not ready to merge) due to an expressed opinion that the technical accuracy and precision aren't currently enough to reflect the importance that Mithril will have to Cardano.

@perturbing can post specifics & in the meantime we'll watch here for developments.  Any review or response from the Mithril team would also be appreciated here (not sure who to tag... maybe @cleanerm5 you could help?).  All editors agreed this is a vital CPS for Cardano and will be ready to keep this document moving towards merge.

rphair reviewed (commented):
@cleanerm5 without your responding yet to any review here, and with no activity at all in 9-10 months (despite work on Cardano light clients proceeding), this CPS itself looks "Abandoned" even through work in the field may still be progressing.

Are you willing to take some time to answer https://github.com/cardano-foundation/CIPs/pull/942#issuecomment-3006920649 in the interest of merging this (as opposed to closing this submission)?  I know it's difficult to respond to a vague objection (not yet clarified by @perturbing here on GitHub) but in the meantime you could indicate some readiness to respond & confirm interest in completing the peer review process.

@perturbing also I think before considering this "Abandoned" we now really need clarification of your objections noted post-meeting and recorded in https://github.com/cardano-foundation/CIPs/pull/942#issuecomment-3071680655.  Authors should be able to expect that objections (e.g. "inadequate for protocols like Mithril") will be documented specifically with text references & constructive suggestions. 🙏

With responses from one or both of you, I feel sure this CIP could be made ready to merge without too much effort.  But without some coordination between author & our most qualified editor to review this, we would have to close it by default.

So to help avoid that default case please @cleanerm5 consider yourself this proposal's one & only advocate, and therefore that the document might not proceed in the review process without your further comment. - cc @polinavino @colll78

rphair commented:
@cleanerm5 I'm closing this because — for reasons above in https://github.com/cardano-foundation/CIPs/pull/942#pullrequestreview-4240354711 — we're not getting the minimum cooperation to produce a peer-reviewed document: even if perhaps the ideas here are compelling or even on the way to implementation.

As this proposal's latest commenters @polinavino @colll78 @nhenin if you would be able to take over work on this proposal — or have any ideas that might conclude the discussion here and make way for a re-submission of a proposal on the same theme — please feel free to comment here: including any requests to re-open this (if so, with an offer to progress & maintain this material).
